Update articles and vb code (#15823)
dotnet/docs: docs/visual-basic/programming-guide/concepts/linq/how-to-query-the-contents-of-files-in-a-folder-linq.md

---
title: "How to query the contents of files in a folder (LINQ) (Visual Basic)"
ms.date: 07/20/2015
ms.assetid: edacbcd3-f3e4-4429-a8be-28a58dc0dd70
---
# How to query the contents of files in a folder (LINQ) (Visual Basic)

This example shows how to query over all the files in a specified directory tree, open each file, and inspect its contents. This type of technique could be used to create indexes or reverse indexes of the contents of a directory tree. A simple string search is performed in this example. However, more complex types of pattern matching can be performed with a regular expression. For more information, see [How to: Combine LINQ Queries with Regular Expressions (Visual Basic)](how-to-combine-linq-queries-with-regular-expressions.md).  
  
## Example  
  
```vb
Imports System.IO

> Imports System.Linq

Reviewer: ```suggestion
```

Module Module1  
    'QueryContents  
    Public Sub Main()  
  
        ' Modify this path as necessary.  

>         Dim startFolder = "c:\program files\Microsoft Visual Studio 9.0\VB\"  

Reviewer: ```suggestion
        Dim startFolder = "C:\Program Files (x86)\Microsoft Visual Studio 14.0"  
```

        ' Take a snapshot of the folder contents.
        Dim dir As New DirectoryInfo(startFolder)
        Dim fileList = dir.GetFiles("*.*", SearchOption.AllDirectories)

>         Dim searchTerm = "Visual Studio"

Reviewer: ```suggestion
        Dim searchTerm = "Welcome"
```

        ' Search the contents of each file.
        ' A regular expression created with the RegEx class
        ' could be used instead of the Contains method.
        Dim queryMatchingFiles = From file In fileList _

>                                  Where file.Extension = ".htm" _

Reviewer: ```suggestion
                                 Where file.Extension = ".html" _
```

                                 Let fileText = GetFileText(file.FullName) _
                                 Where fileText.Contains(searchTerm) _
                                 Select file.FullName

        Console.WriteLine("The term " & searchTerm & " was found in:")

        ' Execute the query.
        For Each filename In queryMatchingFiles
            Console.WriteLine(filename)
        Next

        ' Keep the console window open in debug mode.
        Console.WriteLine("Press any key to exit")
        Console.ReadKey()

    End Sub

    ' Read the contents of the file. This is done in a separate
    ' function in order to handle potential file system errors.
    Function GetFileText(name As String) As String

        ' If the file has been deleted, the right thing
        ' to do in this case is return an empty string.
        Dim fileContents = String.Empty

        ' If the file has been deleted since we took
        ' the snapshot, ignore it and return the empty string.
        If File.Exists(name) Then
            fileContents = File.ReadAllText(name)
        End If

        Return fileContents

    End Function
End Module
```

## Compiling the code

> Create a VB.NET console application project, with an `Imports` statement for the <xref:System.Linq> namespace.

Reviewer: ```suggestion
Create a VB.NET console application project, copy and paste the code sample, and adjust the Startup object value in the project properties.
```
